internal/gengapic/genrest.go

```diff
@@ -0,0 +1,158 @@
+// Copyright (C) 2021  Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gengapic
+
+import (
+	"sort"
+
+	"github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/googleapis/gapic-generator-go/internal/errors"
+	"github.com/googleapis/gapic-generator-go/internal/pbinfo"
+)
+
+func (g *generator) restClientInit(serv *descriptor.ServiceDescriptorProto, servName string, imp pbinfo.ImportSpec, hasRPCForLRO bool) {
+	p := g.printf
+
+	lowcaseServName := lowerFirst(servName)
+
+	p("// Methods, except Close, may be called concurrently. However, fields must not be modified concurrently with method calls.")
+	p("type %sRESTClient struct {", lowcaseServName)
+	p("  host string")
+	p("}")
+	p("")
+	g.restClientUtilities(serv, servName, imp, hasRPCForLRO)
+	g.genRESTMethods(serv, servName)
+}
+
+func (g *generator) genRESTMethods(serv *descriptor.ServiceDescriptorProto, servName string) error {
+	g.addMetadataServiceForTransport(serv.GetName(), "rest", servName)
+
+	methods := append(serv.GetMethod(), g.getMixinMethods()...)
+
+	for _, m := range methods {
+		g.methodDoc(m)
+		if err := g.genRESTMethod(servName, serv, m); err != nil {
+			return errors.E(err, "method: %s", m.GetName())
+		}
+		g.addMetadataMethod(serv.GetName(), "rest", m.GetName())
+	}
+
+	sort.Slice(g.aux.lros, func(i, j int) bool {
+		return g.aux.lros[i].GetName() < g.aux.lros[j].GetName()
+	})
+	for _, m := range g.aux.lros {
+		if err := g.lroType(servName, serv, m); err != nil {
```

Comment:
Would this create duplicate LRO types? For example: `CreateFoo` returns an Operation with a response of `Foo` - the generated `FooOperation` wrapper type would be generated once here and once for gRPC? That would be a collision I believe.

Reply:
You're right. I was punting that issue, but it's better to deal with it now. I'm going to turn `g.aux.lros` into a map set.